util/proxy/dial.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+const DefaultTimeout = 10 * time.Second
+
+// Dialer creates connections using Kubernetes API Server port-forwarding
+type Dialer struct {
+	scope          Scope
+	clientset      *kubernetes.Clientset
+	proxyTransport http.RoundTripper
+	upgrader       spdy.Upgrader
+}
+
+// NewDialer creates a new dialer for a given API server scope
+func NewDialer(s Scope) (*Dialer, error) {
+	if s.Port == 0 {
+		return nil, errors.New("port required")
+	}
+	if s.DialTimeout != nil {
+		s.KubeConfig.Timeout = *s.DialTimeout
+	} else {
+		s.KubeConfig.Timeout = DefaultTimeout
+	}
+	clientset, err := kubernetes.NewForConfig(s.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	proxyTransport, upgrader, err := spdy.RoundTripperFor(s.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &Dialer{
+		scope:          s,
+		proxyTransport: proxyTransport,
+		upgrader:       upgrader,
+		clientset:      clientset,
+	}, nil
+}
+
+// DialContextWithAddr is a GO grpc compliant dialer construct
+func (t Dialer) DialContextWithAddr(ctx context.Context, addr string) (net.Conn, error) {
+	return t.DialContext(ctx, Scheme, addr)
+}
+
+// TODO: Use the context
+// DialContext creates proxied port-forwarded connections
+func (t Dialer) DialContext(ctx context.Context, network string, addr string) (net.Conn, error) {
+	req := t.clientset.CoreV1().RESTClient().
+		Post().
+		Resource(t.scope.Kind).
+		Namespace(t.scope.Namespace).
+		Name(t.scope.ResourceName).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(t.upgrader, &http.Client{Transport: t.proxyTransport}, "POST", req.URL())
+
+	p, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
+	if err != nil {
+		return nil, errors.Wrap(err, "error upgrading connection")
+	}
+	headers := http.Header{}
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	headers.Set(corev1.PortHeader, fmt.Sprintf("%d", t.scope.Port))
+	errorStream, err := p.CreateStream(headers)
```

Comment:
is this as long lived as the `dataStream`? Just curious about how we will resolve the TODO since capturing errors seems pretty important -- Should the errorStream be part of the Dialer struct?

Reply:
gone into the connection